core/src/main/java/org/apache/calcite/interpreter/JoinNode.java

```diff
@@ -47,28 +50,118 @@ public JoinNode(Compiler compiler, Join rel) {
   }
 
   public void run() throws InterruptedException {
-    List<Row> rightList = null;
-    final int leftCount = rel.getLeft().getRowType().getFieldCount();
-    final int rightCount = rel.getRight().getRowType().getFieldCount();
-    context.values = new Object[rel.getRowType().getFieldCount()];
-    Row left;
-    Row right;
-    while ((left = leftSource.receive()) != null) {
-      System.arraycopy(left.getValues(), 0, context.values, 0, leftCount);
-      if (rightList == null) {
-        rightList = new ArrayList<>();
-        while ((right = rightSource.receive()) != null) {
-          rightList.add(right);
+
+    final int fieldCount = rel.getLeft().getRowType().getFieldCount()
+        + rel.getRight().getRowType().getFieldCount();
+    context.values = new Object[fieldCount];
+
+    Source baseSource = leftSource;
```

Comment:
@hsyuan Thanks for reviewing, just a little question.
Might it a little misleading by calling it outer/inner source?
It seems do not match well with  semi/anti/full join.
Any more detail explanation for `outer/inner`?

Reply:
I don't think it is misleading. It is for outer relation or inner relation, not for the outer join or inner join. Look at the nested loop, the outer relation is used for outer loop, the inner relation is used for the inner loop.

Reply:
OK, got your point.
Updated.

Reply:
I think "outer" and "inner" are a bit misleading. Since outer is always left and inner is always right, how about we keep the names left and right? You could add a comment about left being the outer side of the nested loop.

Reply:
But for right outer join, the right relation is outer, left relation is inner.